Feature #147

aleneum opened the pull request:
See #147. Have a look at the [example notebooks](https://github.com/tyarkoni/transitions/tree/feature-%23147/examples) to see it in action.

I went for `show_auto_transitions` to be consistent with `show_conditions` which we already have.
I shortened `show_roi_only` to `show_roi` but can undo that if this is considered less comprehensible.

aleneum commented:
Just fyi:

Coverage decreases because of [this](https://github.com/tyarkoni/transitions/pull/151/files#diff-e29d8b8e9d103e0a5c6a8746a56a0f0cR265) line.
Python optimises some `continue` statements and therefore coverage cannot evaluate it. More info [here](https://bitbucket.org/ned/coveragepy/issues/198/continue-marked-as-not-covered).

wtgee commented on README.md:
```diff
@@ -793,24 +793,30 @@ machine = Machine(model, states, transitions)
 m = Model()
 machine = Machine(model=m, ...)
-m.graph.draw('my_state_diagram.png', prog='dot')
+# Machine(model=m, show_auto_transitions=True, ...)
```
Did you mean to leave this line in here to show the available options? Would be clearer with a comment.

wtgee commented on transitions/extensions/diagrams.py:
```diff
@@ -133,6 +135,7 @@ def get_graph(self, title=None):
         """ Generate a DOT graph with pygraphviz, returns an AGraph object
         Args:
             title (string): Optional title for the graph.
+            show_roi (boolean): Show only the active region if a graph
```
*of* a graph...

aleneum commented:
I used the opportunity to rework the generation of nested graphs. But I am not 100 percent satisfied with how that worked out. The main problem is that Graphviz/dot does not support edges to clusters. In the past, this has not been a big issue since states with children always invoked their first child. I could do the same for clusters and adjust the edges with the attributes `lhead` and `ltail` to hide certain parts of edges. However, since states now can be entered without the need to enter a child state, this becomes more tricky.

I used a hidden anchor state at first which represents the cluster and sticked to `ltail` and `lhead`. This does look messy in certain cases (lower left):


![hide_anchor](https://cloud.githubusercontent.com/assets/205986/19486862/5ee255d6-9560-11e6-9ab2-1d944ac5d77c.png)

In a best case scenario the edges would start and end at the cluster border. But as you see, this does not happen. A transition from the super state to a child state is just floating around and the transition `relax` to state D is very short and the label looks detached.

Thats why I decided to just show the anchor node and its edges:

![show_anchor](https://cloud.githubusercontent.com/assets/205986/19486863/5ee2f2e8-9560-11e6-88f7-d465f14d6964.png)

I'd prefer a solution where edges start and stop at the cluster's border but I could not find a way to achieve this in a generic manner. If you have some thoughts or feedback about how such a graph could be reorganised, let me know.

wtgee commented:
Yeah, it's a little clunky with the anchor but looks a lot better than the free floating version of this. I'll have a chance to test this out on my machines in the next few days but looks good from the overview.

aleneum commented:
Okay,

my next attempt: First, I have split Agraph into `Graph` and `NestedGraph`. This should make both graph generation procedures easier to maintain separately as nested graphs require some specific treatment. I hope this makes `Graph` less 'scary' in cases where someone is trying to understand what it is doing while bug hunting. I also removed the leading 'A' because it may be confused with `pygraphviz.AGraph` otherwise.
Second, I reworked the nested graph layout. The subgraph structure *should* take care of having the anchor black dot at the top of each subgraph. You can get the dot 'source' code with `model.graph.string()`. I copied the output, saved it to a file and tried to work out some decent parameters.
I would not claim that these Graphviz parameters are perfectly adjusted but imho its less cluttered for complex graphs than the previous attempt.
To generate a really awesome looking graph, you probably have to fine tune things anyway.
However, if you find better generic combinations of ranking, spacing and grouping, let me know and I will adapt `NestedGraph`.

Example/notebook graph:
![state2](https://cloud.githubusercontent.com/assets/205986/19627387/1bd62008-9946-11e6-83da-70910b1b27fa.png)

Testing graph:
![state](https://cloud.githubusercontent.com/assets/205986/19627357/5f1ba366-9945-11e6-94d7-ee3a839986bc.png)

aleneum commented:
I could also rework that a bit and only include anchors if there is a transition from parent to child. In this case the result might me a bit more compact. @wtgee, @tyarkoni: What do you think?

tyarkoni commented:
Sorry about the delay. This all looks great to me, and I don't think it's worth the extra effort to make it more compact. I agree that a user who wants something pretty and not just functional should expect to have to customize the graph themselves. Okay to merge as far as I'm concerned.

aleneum commented:
This feature has been integrated into `dev-0.5`
